File: 2015/6/python/main.py
from pathlib import Path
import numpy as np
import re
from typing import Callable, Union, List

def split_at_first_number(s: str) -> tuple[str, str]:
    """Split string into operation and coordinates parts."""
    match = re.search(r'\d', s)
    if not match:
        return s.strip(), ""
    idx = match.start()
    return s[:idx].strip(), s[idx:].strip()

def get_coords(s: str) -> list[list[int, int]]:
    """Parse coordinates string into beginning and end points."""
    if not s:
        return [[0, 0], [0, 0]]
    try:
        data = s.split()
        beginning = list(map(int, data[0].split(",")))
        end = list(map(int, data[-1].split(",")))
        return [beginning, end]
    except (ValueError, IndexError):
        raise ValueError(f"Invalid coordinate format: {s}")

def toggle_cell_part1(grid: np.ndarray, bx: int, by: int, ex: int, ey: int, target: int) -> np.ndarray:
    """Modify grid cells based on operation type for part 1 (binary lights)."""
    # Add 1 to end coordinates to make range inclusive
    ex, ey = ex + 1, ey + 1
    
    # Ensure coordinates are within grid bounds
    bx = np.clip(bx, 0, grid.shape[0])
    by = np.clip(by, 0, grid.shape[1])
    ex = np.clip(ex, 0, grid.shape[0])
    ey = np.clip(ey, 0, grid.shape[1])
    
    if target == 1:  # turn on
        grid[bx:ex, by:ey] = 1
    elif target == 0:  # turn off
        grid[bx:ex, by:ey] = 0
    else:  # toggle
        grid[bx:ex, by:ey] = 1 - grid[bx:ex, by:ey]
    
    return grid

def toggle_cell_part2(grid: np.ndarray, bx: int, by: int, ex: int, ey: int, target: int) -> np.ndarray:
    """Modify grid cells based on operation type for part 2 (brightness levels)."""
    # Add 1 to end coordinates to make range inclusive
    ex, ey = ex + 1, ey + 1
    
    # Ensure coordinates are within grid bounds
    bx = np.clip(bx, 0, grid.shape[0])
    by = np.clip(by, 0, grid.shape[1])
    ex = np.clip(ex, 0, grid.shape[0])
    ey = np.clip(ey, 0, grid.shape[1])
    
    if target == 1:  # turn on: increase by 1
        grid[bx:ex, by:ey] += 1
    elif target == 0:  # turn off: decrease by 1 but not below 0
        grid[bx:ex, by:ey] = np.maximum(0, grid[bx:ex, by:ey] - 1)
    else:  # toggle: increase by 2
        grid[bx:ex, by:ey] += 2
    
    return grid

def process_lights(data: List[str], grid: np.ndarray, toggle_func: Callable) -> int:
    """Process all light operations using the specified toggle function."""
    grid = grid.copy()  # Create a copy to avoid modifying the original
    
    for task in data:
        try:
            operation, coords = split_at_first_number(task)
            beginning, end = get_coords(coords)
            
            bx, by = beginning[0], beginning[1]
            ex, ey = end[0], end[1]
            
            if operation == "turn on":
                grid = toggle_func(grid, bx, by, ex, ey, 1)
            elif operation == "turn off":
                grid = toggle_func(grid, bx, by, ex, ey, 0)
            else:  # toggle operation
                grid = toggle_func(grid, bx, by, ex, ey, -1)
        
        except (ValueError, IndexError) as e:
            print(f"Error processing task '{task}': {e}")
            continue
    
    return int(grid.sum())

def read_input(file_path: Union[str, Path]) -> List[str]:
    """Read and parse input file."""
    try:
        return Path(file_path).read_text().strip().splitlines()
    except FileNotFoundError:
        print(f"Input file not found: {file_path}")
        return []
    except Exception as e:
        print(f"Error reading input file {file_path}: {e}")
        return []

def main() -> None:
    rows, cols = 1000, 1000
    grid = np.zeros((rows, cols))
    
    # Read input files
    data = read_input("../input.txt")
    example = read_input("../example.txt")
    example_2 = read_input("../example_2.txt")
    
    if not data or not example:
        print("Error: Could not process input files")
        return
    
    # Process part 1
    answer_1 = process_lights(data, grid, toggle_cell_part1)
    test_1 = process_lights(example, grid, toggle_cell_part1)
    
    print(f"Part 1 - Main input result: {answer_1}")
    print(f"Part 1 - Example input result: {test_1}")
    
    # Process part 2
    answer_2 = process_lights(data, grid, toggle_cell_part2)
    test_2 = process_lights(example_2, grid, toggle_cell_part2)
    
    print(f"Part 2 - Main input result: {answer_2}")
    print(f"Part 2 - Example input result: {test_2}")

if __name__ == "__main__":
    main()

File: 2015/6/go/main.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func splitAtFirstNumber(s string) (string, string) {
	re := regexp.MustCompile(`\d`)
	match := re.FindStringIndex(s)

	if match == nil {
		return strings.TrimSpace(s), ""
	}

	idx := match[0]
	return strings.TrimSpace(s[:idx]), strings.TrimSpace(s[idx:])
}

func getCoords(s string) ([][]int, error) {
	if s == "" {
		return [][]int{{0, 0}, {0, 0}}, nil
	}

	parts := strings.Split(s, " ")
	if len(parts) < 1 {
		return nil, fmt.Errorf("invalid coordinate format: %s", s)
	}

	beginningParts := strings.Split(parts[0], ",")
	endParts := strings.Split(parts[len(parts)-1], ",")

	beginning := make([]int, 2)
	end := make([]int, 2)

	var err error
	beginning[0], err = strconv.Atoi(beginningParts[0])
	if err != nil {
		return nil, err
	}
	beginning[1], err = strconv.Atoi(beginningParts[1])
	if err != nil {
		return nil, err
	}

	end[0], err = strconv.Atoi(endParts[0])
	if err != nil {
		return nil, err
	}
	end[1], err = strconv.Atoi(endParts[1])
	if err != nil {
		return nil, err
	}

	return [][]int{beginning, end}, nil
}

func toggleCellPart1(grid [][]int, bx, by, ex, ey, target int) [][]int {
	rows, cols := len(grid), len(grid[0])

	// Clip coordinates
	if bx < 0 {
		bx = 0
	}
	if by < 0 {
		by = 0
	}
	if ex > rows {
		ex = rows
	}
	if ey > cols {
		ey = cols
	}

	for x := bx; x < ex; x++ {
		for y := by; y < ey; y++ {
			switch target {
			case 1: // turn on
				grid[x][y] = 1
			case 0: // turn off
				grid[x][y] = 0
			default: // toggle
				grid[x][y] = 1 - grid[x][y]
			}
		}
	}

	return grid
}

func toggleCellPart2(grid [][]int, bx, by, ex, ey, target int) [][]int {
	rows, cols := len(grid), len(grid[0])

	// Clip coordinates
	if bx < 0 {
		bx = 0
	}
	if by < 0 {
		by = 0
	}
	if ex > rows {
		ex = rows
	}
	if ey > cols {
		ey = cols
	}

	for x := bx; x < ex; x++ {
		for y := by; y < ey; y++ {
			switch target {
			case 1: // turn on: increase by 1
				grid[x][y]++
			case 0: // turn off: decrease by 1 but not below 0
				if grid[x][y] > 0 {
					grid[x][y]--
				}
			default: // toggle: increase by 2
				grid[x][y] += 2
			}
		}
	}

	return grid
}

func processLights(data []string, grid [][]int, toggleFunc func([][]int, int, int, int, int, int) [][]int) int {
	gridCopy := make([][]int, len(grid))
	for i := range grid {
		gridCopy[i] = make([]int, len(grid[0]))
		copy(gridCopy[i], grid[i])
	}

	for _, task := range data {
		operation, coords := splitAtFirstNumber(task)
		coordsParts, err := getCoords(coords)
		if err != nil {
			fmt.Printf("Error processing task '%s': %v\n", task, err)
			continue
		}

		bx, by := coordsParts[0][0], coordsParts[0][1]
		ex, ey := coordsParts[1][0]+1, coordsParts[1][1]+1

		switch operation {
		case "turn on":
			gridCopy = toggleFunc(gridCopy, bx, by, ex, ey, 1)
		case "turn off":
			gridCopy = toggleFunc(gridCopy, bx, by, ex, ey, 0)
		case "toggle":
			gridCopy = toggleFunc(gridCopy, bx, by, ex, ey, -1)
		}
	}

	return sumGrid(gridCopy)
}

func sumGrid(grid [][]int) int {
	total := 0
	for _, row := range grid {
		for _, val := range row {
			total += val
		}
	}
	return total
}

func readInput(filePath string) []string {
	content, err := os.ReadFile(filePath)
	if err != nil {
		fmt.Printf("Error reading input file %s: %v\n", filePath, err)
		return []string{}
	}

	lines := strings.Split(strings.TrimSpace(string(content)), "\n")
	return lines
}

func main() {
	rows, cols := 1000, 1000
	grid := make([][]int, rows)
	for i := range grid {
		grid[i] = make([]int, cols)
	}
	data := readInput("../input.txt")

	if len(data) == 0 {
		fmt.Println("Error: Could not process input files")
		return
	}
	answer1 := processLights(data, grid, toggleCellPart1)

	fmt.Printf("Part 1 - Main input result: %d\n", answer1)
	answer2 := processLights(data, grid, toggleCellPart2)

	fmt.Printf("Part 2 - Main input result: %d\n", answer2)

}
